include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -410,10 +410,21 @@ class transform_iterator
     typedef typename ::std::iterator_traits<_Iter>::pointer pointer;
     typedef typename ::std::iterator_traits<_Iter>::iterator_category iterator_category;
 
-    transform_iterator(_Iter __it = _Iter(), _UnaryFunc __unary_func = _UnaryFunc())
-        : __my_it_(__it), __my_unary_func_(__unary_func)
+    //only enable this constructor if both template types are default constructible
+    template <typename _Iter_ = _Iter, typename _UnaryFunc_ = _UnaryFunc,
+              typename = ::std::enable_if_t<
+                  ::std::is_default_constructible_v<_Iter_>&& ::std::is_default_constructible_v<_UnaryFunc_>>>
+    transform_iterator() : __my_it_(), __my_unary_func_()
```

Comment:
`= default` should be enough

Reply:
Yes, I agree it is better for many reasons.

Reply:
Actually, `= default` causes problems in combination with the `const` unary functor.  It implicitly deletes the default constructor because of this const member variable.  In the future, we may be removing `const` from the unary functor, but lets have that discussion separately, in the context of potentially removing the custom assignment operator. 

For now, we can use this version with `enable_if`.  We lose default constructability and the code is ugly, but I think its the best option for now.

Reply:
It looks weird that for `const` member it actually deletes the default constructor...

Reply:
If we want to have our code cleaner and keep `const` (purpose of which I don't understand) we can to like that:

```cpp
const _UnaryFunc __my_unary_func_ = _UnaryFunc();

transform_iterator() = default;
```

We lose default constructor triviality anyway. So, seems like for `const` member trivial default constructor may not exist. That makes it more evil than value

Reply:
My best understanding of why the functor is `const` here is to convey to users that the functor in a `transform_iterator` should be stateless, and should not be relied upon to collect some data as it operates to affect the transformation, or for later retrieval.    

In the long run, I do believe we should remove `const`.  

Sadly, I think your suggestion to use `const _UnaryFunc __my_unary_func_ = _UnaryFunc();` results in something similar to the original issue, a build error when `is_default_constructible` is called on clang when `_UnaryFunc` is non-default constructible.

Reply:
Ok, I see. My experiments were not quite correct. What is we remove `const` right now? Does it break somebody's code?

Reply:
I don't think it can break any code to remove it, but I have not investigated this in detail.  I know @akukanov had mentioned trying to find some other way to convey the same message to users of `transform_iterator` if we decide to remove `const`.  

Perhaps that just comes in the form of documentation. I would be open to removing `const` and adding some language to the documentation.

Reply:
The thing is our `operator*` is const anyway. So having some state within lambda and trying to call mutable `operator()` ends up with [compile-time error](https://godbolt.org/z/M17obYqh3). So, I don't know how much value we bring to the users by marking the member of transform iterator as `const`. Perhaps there is a value in `operator=` but so far I am hearing that it rather creates problems for us (if my understanding is correct)

Reply:
I'm in agreement that `const` should be removed.  We can go ahead and do it here, then use the defaulted ctor.  I'll look at the documentation to see if there is a good way to improve it to make the user understand how transform iterator should be used, and what functors are eligible.

Reply:
I've removed `const` and switched back to `transform_iterator() = default` .

I've added a check `__is_const_callable_with_arg_v` to the definition of `transform_iterator` to ensure that a const version of the functor is callable with the reference type of the base iterator.  This check needs review, because if it is more aggressive than the intention, then this could cause problems for user code.

I've also added some to the documentation of `tranform_iterator`.